// replicated-config-render/lib/components/ConfigFieldFile.js
'use strict';

var React = require('react');
var ReactDOM = require('react-dom');

var FileInput = require('./FileInput');

var ConfigFieldFile = React.createClass({
  displayName: 'ConfigFieldFile',

  handleOnChange: function handleOnChange(value, data) {
    ReactDOM.findDOMNode(this.refs.file).value = '';
    if (this.props.handleChange) {
      if (this.props.multiple) {
        this.props.handleChange(this.props.name, value, data);
      } else {
        this.props.handleChange(this.props.name, value ? value[0] : '', data ? data[0] : '');
      }
    }
  },

  getFilenamesText: function getFilenamesText() {
    if (this.props.multiple) {
      if (this.props.multi_value && this.props.multi_value.length) {
        return this.props.multi_value.join(', ');
      }
    } else if (this.props.value) {
      return this.props.value;
    }
    return this.props['default'];
  },

  render: function render() {
    return React.createElement(
      'div',
      { className: 'field field-type-file' },
      React.createElement(
        'label',
        null,
        this.props.title
      ),
      React.createElement(
        'div',
        { className: 'input input-type-file clearfix' },
        React.createElement(
          'div',
          null,
          React.createElement(
            'span',
            null,
            React.createElement(FileInput, {
              ref: 'file',
              name: this.props.name,
              readOnly: this.props.readonly,
              multiple: this.props.multiple,
              onChange: this.handleOnChange })
          ),
          React.createElement(
            'small',
            null,
            this.getFilenamesText()
          )
        )
      )
    );
  }
});

module.exports = ConfigFieldFile;
