// replicated-config-render/lib/components/ConfigFieldText.js
'use strict';

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var classNames = require('classnames');
var React = require('react');

var ConfigFieldText = React.createClass({
  displayName: 'ConfigFieldText',

  handleOnChange: function handleOnChange(ev) {
    if (this.props.handleChange) {
      this.props.handleChange(this.props.name, ev.target.value);
    }
  },

  render: function render() {
    var inputClasses = classNames('form-control', {
      'readonly': this.props.readOnly
    });

    return React.createElement(
      'div',
      { className: 'field field-type-text' },
      React.createElement(
        'label',
        null,
        this.props.title
      ),
      React.createElement(
        'div',
        { className: 'input input-type-text' },
        React.createElement('input', _extends({
          type: 'text'
        }, this.props.props, {
          placeholder: this.props['default'],
          value: this.props.value,
          readOnly: this.props.readOnly,
          onChange: this.handleOnChange,
          className: inputClasses }))
      )
    );
  }
});

module.exports = ConfigFieldText;
