// replicated-config-render/lib/components/ConfigFieldRadio.js
'use strict';

var _ = require('lodash');
var classNames = require('classnames');
var React = require('react');
var ReactDOM = require('react-dom');

var ConfigFieldRadio = React.createClass({
  displayName: 'ConfigFieldRadio',

  handleOnClick: function handleOnClick() {
    var node = ReactDOM.findDOMNode(this.refs.radio);
    if (!node.checked) {
      node.checked = true;
      this.handleOnChange();
    }
  },

  handleOnChange: function handleOnChange() {
    var val = ReactDOM.findDOMNode(this.refs.radio).checked ? '1' : '0';
    if (val === '1' && this.props.handleChange) {
      this.props.handleChange(this.props.name, val);
    }
  },

  render: function render() {
    var val = _.get(this.props, 'value', null);
    if (!val || val.length === 0) {
      val = this.props['default'];
    }
    var checked = val === this.props.name;

    var inputClasses = classNames('btn', 'btn-default', 'btn-checkbox', {
      'active': checked,
      'readonly': this.props.readOnly
    });

    var iconClasses = classNames('fa', {
      'fa-dot-circle-o': checked,
      'fa-circle-o': !checked
    });

    return React.createElement(
      'div',
      { className: 'field field-type-radio' },
      React.createElement(
        'div',
        { className: 'input input-type-radio' },
        React.createElement(
          'button',
          {
            type: 'button',
            disabled: this.props.readOnly,
            onClick: this.handleOnClick,
            className: inputClasses },
          React.createElement('i', { className: iconClasses }),
          ' ',
          this.props.title
        ),
        React.createElement('input', {
          ref: 'radio',
          type: 'radio',
          name: this.props.group,
          value: this.props.name,
          checked: checked,
          onChange: this.handleOnChange,
          className: 'hidden' })
      )
    );
  }
});

module.exports = ConfigFieldRadio;
