// replicated-config-render/lib/components/Markdown.js
'use strict';

var _ = require('lodash');
var marked = require('marked');
var classNames = require('classnames');
var React = require('react');
var ReactDOM = require('react-dom');

var Markdown = React.createClass({
  displayName: 'Markdown',

  componentDidMount: function componentDidMount() {
    this.openLinksInNewWindow();
  },

  componentDidUpdate: function componentDidUpdate() {
    this.openLinksInNewWindow();
  },

  openLinksInNewWindow: function openLinksInNewWindow() {
    _.each(ReactDOM.findDOMNode(this.refs.markdown).getElementsByTagName('a'), function (el) {
      el.setAttribute('target', '_blank');
    });
  },

  render: function render() {
    var classes = classNames('markdown', this.props.className);
    var md = '';
    var err = null;
    if (this.props.value) {
      try {
        md = marked(this.props.value);
      } catch (exc) {
        err = exc.toString();
      }
    }
    return React.createElement(
      'div',
      null,
      React.createElement('div', { ref: 'markdown', className: classes,
        dangerouslySetInnerHTML: { __html: md } }),
      React.createElement(
        'div',
        { className: 'text-danger' },
        err
      )
    );
  }
});

module.exports = Markdown;
